### PR TITLE
BT-480 - migration workflow

### PR DESCRIPTION
**Pull request by mattoberle**
This PR adds several improvements to the NYPR setuptools extensions.

1. Tests are added for DockerDeploy commands.
2. A migration workflow is now available through the `--migrate` flag.
3. The DockerDeploy class is re-organized to provide better separation between operations.

To run migrations a flag like `--migrate='./manage.py migrate'` can be provided.
When paired with other deployment flags the workflow looks like this:

1. Build Docker image.
2. (Optional) Test Docker image.
3. Push Docker image.
4. Update ECS Task Definition.
5. Run migrations by launching a single ECS task.
6. Update ECS Service.

### CODE REVIEW

File: nyprsetuptools/commands/deploy.py [hunk 1]
```diff
@@ -188,63 +209,166 @@ def run(self):
...
+            'count': 1,
+        }
+        if self.fargate:
+            service = self.ecs.describe_services(
+                cluster=cluster_name,
+                services=[task_name],
+            )['services'][0]
+            network_config = service['networkConfiguration']
+            options['launchType'] = 'FARGATE'
+            options['networkConfiguration'] = network_config
```
[the remark this answers is not in the document]
Comment (mattoberle) [reply nyprsetuptools/commands/deploy.py]: Yeah networkConfiguration is required for Fargate (to specify VPC, subnet, etc).
When running on EC2 the EC2 instance's network is used.

File: nyprsetuptools/commands/deploy.py [hunk 2]
```diff
@@ -188,63 +209,166 @@ def run(self):
...
+        migration_task = wait(self.ecs.run_task, kwargs=options,
+                              wait_for_func=lambda x: x['tasks'])
+
+        uuid = migration_task['tasks'][0]['taskArn'].split('/')[-1]
+        resp = self.ecs.describe_task_definition(
+            taskDefinition=task_name
+        )
+        log_config = resp['taskDefinition']['containerDefinitions'][0]['logConfiguration']['options']
+        log_group_name = log_config['awslogs-group']
+        log_stream_name = '{prefix}/{family}/{uuid}'.format(
```
[the remark this answers is not in the document]
Comment (mattoberle) [reply nyprsetuptools/commands/deploy.py]: Hmmm, great question.

[the remark this answers is not in the document]
Comment (ncastro-nypr) [reply nyprsetuptools/commands/deploy.py]: if it works as is, there's no reason to change it, I was just curious. A cursory inspection of the log stream names indicated that they're always log_group_name+uuid but I didn't really dig into it much. It's not important haha.

File: nyprsetuptools/commands/deploy.py [hunk 3]
```diff
@@ -188,63 +209,166 @@ def run(self):
...
+            resp = wait(self.cwl.get_log_events, kwargs=options,
+                        exceptions=[self.ClientError],
+                        wait_for_func=lambda x: x['events'])
+            if not resp:
+                return
+            for event in resp['events']:
+                print(event['message'])
+            else:
+                return
+            read_logs(resp['nextForwardToken'])
```
[the remark this answers is not in the document]
Comment (mattoberle) [reply nyprsetuptools/commands/deploy.py]: This is the part I need to touch up.
Line 275 *does* get called in the case where `resp['events']` is populated with entries.
Each entry in the log will be printed and then `read_logs` will get called recursively with the `nextForwardToken`.

Line 275 will *not* get called if no response is received from `self.cwl.get_log_events` or if no events are returned in the response.

File: nyprsetuptools/commands/deploy.py [hunk 4]
```diff
@@ -188,63 +209,166 @@ def run(self):
...
+            # of the latest task definition, however this script is
+            # unaware of the previous task definition ARN.
+            # The key not matching the new ARN will be retrieved as 'old'.
+            deployments = {d['taskDefinition']: d
+                           for d in resp['services'][0]['deployments']}
+            new = deployments.pop(task_definition_arn)
+            if deployments:
+                old = deployments.pop(list(deployments.keys())[0])
+            else:
+                old = {'runningConut': 0}
```
[the remark this answers is not in the document]
Comment (mattoberle) [reply nyprsetuptools/commands/deploy.py]: WOW. This is why deployments always looked like they failed for new services. Thank you!

[the remark this answers is not in the document]
Comment (ncastro-nypr) [reply nyprsetuptools/commands/deploy.py]: yeah I figured there might be some weirdness like that. it's always the little things :D

**Pull request merged**